javatests/arcs/core/analysis/EntityTypeUnionTest.kt

```diff
@@ -0,0 +1,100 @@
+package arcs.core.analysis
+
+import arcs.core.data.EntityType
+import arcs.core.data.FieldType
+import arcs.core.data.Schema
+import arcs.core.data.SchemaFields
+import arcs.core.data.SchemaName
+import com.google.common.truth.Truth.assertThat
+import org.junit.Assert.assertTrue
+import org.junit.Assert.assertNotNull
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.junit.runners.JUnit4
+
+@RunWith(JUnit4::class)
+class EntityTypeUnionTest {
+    // TODO(bgogul): hash for all Schema instances.
+
+    @Test
+    fun schemaUnionMergesNames() {
+        val emptySchemaFields = SchemaFields(emptyMap(), emptyMap())
+        val thingSchema = Schema(
+            setOf(SchemaName("Thing"), SchemaName("Another")),
+            emptySchemaFields,
+            ""
+        )
+        val objectSchema = Schema(
+            setOf(SchemaName("Object"), SchemaName("Thing")),
+            emptySchemaFields,
+            ""
+        )
+        val thingObjectSchema = thingSchema.union(objectSchema).getOrNull()
+        assertThat(thingObjectSchema).isNotNull()
+        thingObjectSchema?.let {
+            assertThat(it.names).containsExactly(
+                SchemaName("Thing"),
+                SchemaName("Object"),
+                SchemaName("Another")
+            )
+        }
+    }
+
+    @Test
+    fun schemaUnionMergesSingletons() {
+        val textField = SchemaFields(
+            mapOf("text" to FieldType.Text),
+            emptyMap()
+        )
+        val numberField = SchemaFields(
+            mapOf("number" to FieldType.Number),
+            emptyMap()
+        )
+        val textSchema = Schema(setOf(SchemaName("Example")), textField, "")
+        val numberSchema = Schema(setOf(SchemaName("Example")), numberField, "")
+        val result = textSchema.union(numberSchema).getOrNull()
+        result?.let {
```

Comment:
This will skip running the assertions below if result is null, which is not what you want to happen.

You have a few options here:

```
val result = textSchema.union(numberSchema).getOrNull()!!
val result = requireNotNull(textSchema.union(numberSchema).getOrNull())
val result = textSchema.union(numberSchema).getOrThrow()
```

(`getOrThrow` you'd have to write. In Rust they call that function `unwrap`, but I don't know if that's a common enough term that everyone would understand it)

Reply:
I meant to add a `assertThat(result).isNotNull()` before this call, but seems to have missed it in this case. Added it now.

Reply:
This construction is still pretty dangerous, since forgetting the `isNotNull` call can cause it to completely skip all of the assertions, whereas with the other options I suggested, forgetting to use them causes a compile error (nullability).

I'd recommend `!!` as the very easiest option.

Reply:
Ah, makes sense. Used `requireNotNull` as it is more readable.

Reply:
Sweet thanks! Now there's no need for the `result?.let` at all

Reply:
Good point! removed `?.`.